### PR TITLE
do not merge: Variant support in components

### LINKED ISSUE #14669: Variants PR reviewed and checked in
Variants PR reviewed and checked in

### PR DESCRIPTION
**Pull request by dzearing**
This is an initial PR to support component variants using hooks; not the final form but wanted to get perf numbers earlier to see what needs tweaking. Fixes #14669

## Variants support

When building components, the component will come with a default stylesheet, piping in css variables (tokens) into the right place.

Each token may have a default value. Modifiers on the component may change the default values of those tokens. And, users will want to adjust those defaults through theming.

Variants are defined as such:

### 1. Create the variant hook 

Note this is created outside of the component scope.

```jsx
const useFooClasses = makeVariantClasses({
  name: 'Foo',
  prefix: '--foo', 
  styles: {
    root: { ... },
    icons: { ... },
  },
  variants: {
    root: { 
      /* baseline default token values */ 
      background: 'red'
    },
    primary: { 
      /* ...tokens which apply when the primary flag is set or the `variant` name on the component is "primary" */ 
      background: 'green'
    }
  }
});
```

### 2. The hook is used in a component

```jsx
const Foo = (props) => {
  const state = { ...props };

  useFooClasses (state);

  return (
    <div className={state.className} />
  );
```

### 3. The component can be used by the customer 

```jsx
<Foo /> // red foo
<Foo primary /> // green foo
<Foo variant="primary" /> // variant string prop can also be used
```

### 4. The component's variants can be themed

```jsx
<ThemeProvider theme={{
  components: {
     Foo: {
       variants: {
          primary: {
             background: 'purple'
          }
        }
      }
    }
  }}>
  <Foo/> // red button
  <Foo primary /> // purple button
</ThemeProvider>
```

### CODE REVIEW

File: packages/react-theme-provider/src/getTokens.ts [hunk 1]
```diff
@@ -68,168 +68,8 @@ export function getTokens(theme: Theme): Tokens {
           },
         },
       },
-
-      body: {
```
Comment (dzearing) [inline packages/react-theme-provider/src/getTokens.ts:72]: dont remove

File: packages/react-button/src/components/Button/Button.types.ts [hunk 2]
```diff
@@ -89,61 +91,62 @@ export type ButtonProps = ComponentProps &
...
 
-export interface ButtonSlots extends BaseSlots {
-  icon: React.ElementType;
-  loader: React.ElementType;
-}
-
-export type ButtonSlotProps = SlotProps<ButtonSlots, ButtonProps, React.ButtonHTMLAttributes<HTMLButtonElement>>;
-
-export type ButtonTokenSet = ColorTokenSet & {
+export type ButtonTokens = ColorTokens & {
```
Comment (xugao) [inline packages/react-button/src/components/Button/Button.types.ts:101]: typing doesn't look right; if you compare this with what we have in `useButtonVariants`

File: packages/react-button/src/components/Button/useButtonClasses.ts [hunk 3]
```diff
@@ -137,79 +137,6 @@ export const useButtonClasses = makeClasses<ButtonState>({
...
-  _circular: {
-    '--button-borderRadius': '50000px',
-  },
-
-  _fluid: {
-    '--button-width': '100%',
-    '--button-maxWidth': '100%',
-  },
-
   _size_smallest: {
```
Comment (xugao) [inline packages/react-button/src/components/Button/useButtonClasses.ts:140]: we should move these too?

File: packages/react-button/src/components/Button/useButtonVariants.ts [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        smaller: '24px',
+        small: '24px',
+        regular: '32px',
+        large: '40px',
+        larger: '48px',
+        largest: '64px',
+      },
+
+      // Sizing tokens
+      iconSize: fonts?.mediumPlus?.fontSize,
```
Comment (xugao) [inline packages/react-button/src/components/Button/useButtonVariants.ts:20]:
```suggestion
      iconSize: fonts.mediumPlus.fontSize,
```

we should not need `?` everywhere

File: packages/react-theme-provider/src/makeVariants.test.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            .first()
+            .getDOMNode()
+            .getAttribute('class'),
+        ).toEqual('Button-0');
+        expect(_stylesheet.getRules()).toEqual('.Button-0{--button-background:red;}');
+      },
+    );
+  });
+
+  it('can render primary variant', () => {
```
Comment (xugao) [inline packages/react-theme-provider/src/makeVariants.test.tsx:61]:
```suggestion
  it('can render modifier variant', () => {
```

File: packages/react-theme-provider/src/makeVariants.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  state: GenericDictionary,
+  variantKeys: string[],
+  variantObjects: TokenSetType[],
+) => {
+  if (variants) {
+    for (const variantName of Object.keys(variants)) {
+      const isBase = variantName === 'base';
+      if (variants[variantName] && (isBase || state[variantName] || state.variant === variantName)) {
+        variantKeys.push(isBase ? componentName : variantName);
+        variantObjects.push(variants[variantName]!);
```
Comment (xugao) [inline packages/react-theme-provider/src/makeVariants.ts:28]:
```suggestion
        variantObjects.push(variants[variantName]);
```

File: packages/react-theme-provider/etc/react-theme-provider.api.md [hunk 7]
```diff
@@ -1,151 +1,157 @@
...
+export const createDefaultTheme: () => Theme;
+
+// @public (undocumented)
+export const FluentTheme: Theme;
+
+// @public (undocumented)
+export type FontFace = IFontFace;
+
+// @public (undocumented)
+export type GenericDictionary = Record<string, any>;
```
Comment (xugao) [inline packages/react-theme-provider/etc/react-theme-provider.api.md:30]: i don't think we should export this

File: packages/react-button/src/components/Button/useButtonVariants.ts (hunk 4 above)
Comment (xugao) [reply packages/react-button/src/components/Button/useButtonVariants.ts:20]: pls sync code with latest code from `getTokens`

File: packages/theme/src/types/Theme.ts [hunk 8]
```diff
@@ -4,18 +4,18 @@ import { IStyleFunctionOrObject } from '@uifabric/utilities';
...
-  secondaryContentColor: string;
-  linkColor: string;
-  iconColor: string;
-  borderColor: string;
-  dividerColor: string;
-  focusColor: string;
-  focusInnerColor: string;
-  opacity: string;
-}>;
+export type ColorTokenSet = {
```
Comment (xugao) [inline packages/theme/src/types/Theme.ts:7]: why not using `Partial`?

File: packages/theme/src/types/Theme.ts [hunk 9]
```diff
@@ -37,16 +37,14 @@ export type ColorTokens = Partial<{
...
-}>;
-
-export type ColorTokenSet = ColorTokens & ColorTokenStates;
+export type ColorTokens = ColorTokenSet & {
+  hovered?: ColorTokens;
+  pressed?: ColorTokens;
+  disabled?: ColorTokens;
+  checked?: ColorTokens;
+  checkedHovered?: ColorTokens;
+  checkedPressed?: ColorTokens;
```
Comment (xugao) [inline packages/theme/src/types/Theme.ts:46]: why not using `Partial` (what we had before)?

File: packages/theme/etc/theme.api.md [hunk 10]
```diff
@@ -729,6 +730,9 @@ export type TokenSetType = {
     [key: string]: TokenSetType | string | number | undefined;
 };
 
+// @public (undocumented)
+export type Variants = Record<string, any>;
```
Comment (xugao) [inline packages/theme/etc/theme.api.md:734]: Variants should always contain base, something like:

```
interface Variants {
  base?: RecursivePartial<Tokens>;
  [key: string]: RecursivePartial<Tokens>;
} 
```

File: packages/react-theme-provider/src/makeVariants.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+ */
+export const makeVariants = <TTokenSetType extends TokenSetType>(
+  componentName: string,
+  prefix: string,
+  variants: Variants | ((theme: Theme) => Variants),
+) => {
+  const variantToClassName: Record<string, string> = {}; // useThemeSettings(id);
+
+  return (state: GenericDictionary) => {
+    // Grab the theme.
```
[the remark this answers is not in the document]
Comment (dzearing) [reply packages/react-theme-provider/src/makeVariants.ts]: done

File: packages/react-button/src/components/Button/Button.types.ts [hunk 12]
```diff
@@ -89,61 +91,62 @@ export type ButtonProps = ComponentProps &
...
+  borderWidth?: string;
+  boxShadow?: string;
+  width?: string;
+
+  transform?: string;
+  transition?: string;
+
+  fontFamily?: string;
+  fontSize?: string;
+  fontWeight?: string;
```
[the remark this answers is not in the document]
Comment (dzearing) [reply packages/react-button/src/components/Button/Button.types.ts]: done

File: packages/react-button/src/components/Button/useButtonClasses.ts (hunk 3 above)
Comment (dzearing) [reply packages/react-button/src/components/Button/useButtonClasses.ts]: done.

File: packages/theme/src/types/Theme.ts (hunk 8 above)
Comment (dzearing) [reply packages/theme/src/types/Theme.ts:12]: It seemed more ergonomic.

File: packages/theme/etc/theme.api.md (hunk 10 above)
Comment (dzearing) [reply packages/theme/etc/theme.api.md:859]: I think this we should leave as is. I don't think base (changed to root) is required.

File: packages/react-theme-provider/src/makeVariants.ts (hunk 6 above)
Comment (dzearing) [reply packages/react-theme-provider/src/makeVariants.ts]: no longer applicable.

File: .vscode/settings.json [hunk 13]
```diff
@@ -15,7 +15,7 @@
     "**/.git": true,
     "**/.DS_Store": true,
     "**/coverage": true,
-    "**/*.scss.ts": true,
+    "**/*.scss.ts": false,
```
Comment (dzearing) [inline .vscode/settings.json:18]: revert

File: .vscode/launch.json [hunk 14]
```diff
@@ -14,6 +14,7 @@
       "env": {
         "NODE_ENV": "development"
       },
+      "outFiles": ["${workspaceFolder}/lib-commonjs/**/*"],
```
Comment (dzearing) [inline .vscode/launch.json:17]: revert

**Pull request closed without merging**